_posts/pl/2021-02-10-przyklad-migracji-do-chmury.md

```diff
@@ -0,0 +1,888 @@
+
+W systemie docelowym, jako że jest on w infrastrukturze chmury (czyli dostępny publicznie przez Internet), w celu zabezpieczenia przed nieupoważnionym dostępem, musimy zaimplementować mechanizm uwierzytelniania (chyba, że nie obawiamy się rachunków od Google). W tym celu wykorzystamy [Google Sign-In](https://developers.google.com/identity). Aby to zrobić, musimy w [APIs & Services](https://console.cloud.google.com/apis/credentials) stworzyć credential **OAuth client ID** i skonfigurować w Cloud IAM przykładowego użytkownika, który posiadać będzie uprawnienia do korzystania z aplikacji (może to być nasze główne konto, które posiada rolę właściciela).
+
+Dzięki wykorzystaniu Cloud Functions, zoptymalizujemy koszt utrzymania systemu. Funkcje generują koszt wyłącznie wtedy, kiedy działają - za liczbę wywołań oraz czas ich wykonywania. W przypadku bardzo niskiego ruchu (łapiącego się w darmowe limity wywołań funkcji), lub jego braku, opłaty mogą zostać zminimalizowane do zera. Funkcje skalują się automatycznie, dlatego cała odpowiedzialność za obsłużenie ruchu użytkowników i zapewnienie dostępności systemu leży po stronie usługodawcy.
+
+Usługa Cloud Pub/Sub działa na zasadzie kolejki, więc idealnie zastąpi RabbitMQ. Zazwyczaj Cloud Pub/Sub dostarcza każdą wiadomość jeden raz oraz w takiej kolejności, w jakiej została opublikowana. Google w dokumentacji usługi informuje, że mogą zdarzyć się sytuacje, w których wiadomość zostanie dostarczona poza kolejnością lub więcej niż jeden raz. Wiedząc o tym, należy zaimplementować odbiorcę wiadomości w sposób idempotentny, czyli odporny na wielokrotne dostarczenie tej samej wiadomości. W przypadku naszej aplikacji demo nie stanowi to problemu, ponieważ każda wiadomość posiadać będzie unikalny identyfikator, a wielokrotne zapisanie tej samej wiadomości do Cloud Firestore spowoduje jej nadpisanie.
+
+Pojawienie się wiadomości w Cloud Pub/Sub triggerować będzie wywołanie funkcji odpowiedzialnej za przetworzenie wiadomości i utrwalenie jej w Cloud Firestore. Zrezygnujemy z funkcji pomiędzy aplikacją frontendową a Cloud Pub/Sub, ponieważ zakładamy, że nie ma potrzeby walidacji danych przed wysłaniem ich do Cloud Pub/Sub. W aplikacji źródłowej taka walidacja była wykonywana przez aplikację backendową, zanim wiadomość została wysłana do kolejki RabbitMQ. Dzięki temu, podczas migracji zoptymalizujemy system, upraszczając jego architekturę. Oczywiście taka optymalizacja byłaby możliwa również w systemie źródłowym, ponieważ RabbitMQ umożliwia dostarczanie wiadomości poprzez REST API, wykorzystując w tym celu dodatkową wtyczkę.
+
+Odpowiednikiem bazy MongoDB jest usługa Cloud Firestore. Umożliwia ona zapis i odczyt za pomocą REST API, dzięki czemu odczyt może być wykonywany w samej aplikacji frontendowej, bez udziału aplikacji backendowej (nie jest potrzebny sterownik do bazy danych). To upraszcza jeszcze bardziej architekturę systemu docelowego.
```

Comment:
no upraszczamy i to bardzo, ale kosztem transakcji. system przed migracja nie gubił wiadomości pomiędzy kolejką a bazą, a ten na chmurce już tak

Reply:
Co masz na myśli pisząc, że system docelowy gubi wiadomości pomiędzy Pub/Sub a Firestore? Chodzi o brak ponawiania? Nie implementowałem go, wspominając jedynie o tym, ponieważ czekam na wpis na bloga o samej konfiguracji Pub/Sub (ponawianie + DLQ) :)

Reply:
Chodzi mi o to, że jak masz Listenera, który jest uwieszony na kolejce na Rabbicie i zapisuje to mongo i ten zapis się nie uda, to wiadomość wraca do kolejki. Masz to domyślnie z racji tego, że sterownik do Rabbita działa w transakcji. Jak przeniesiesz to na Pub/Sub i Firestore po REST, no to nie masz juz tego mechanizmu.

Reply:
Tylko tutaj pojawienie się wiadomości na Pub/Sub triggeruje funkcję, która kończy się błędem w przypadku jeśli nie uda się zapisać do Firestore. Więc wtedy taka wiadomość nie dostanie ACK i powinna zostać ponowiona. Było to widać jak z Markiem zrobiliśmy błąd w payoffsach i funkcja napieprzała w kółko, bo za każdym razem kończyła się błędem.

Reply:
No tak ale to zadziała tylko wtedy, gdy poprawnie obsłużysz odpowiedz z REST API Firestora. Jeśli zignorujesz kod błędu http to wiadomość znika. I taką sytuację też mieliśmy w payoffs :) Podsumowując, piszesz tutaj o upraszczaniu ale w rzeczywistości ten kod transakcyjny jest prosty. Bo tam trwałość danych gwarantują ci transakcje i sterowniki a w chmure musisz sam sobie to oprogramować i skonfigurować odpowiednio, żeby uzyskać ten sam efekt.

Reply:
Zgadza się :) Nie implementowałem listenera dla Rabbita, który zapisuje do bazy, więc nawet nie wiem jak to wygląda w praktyce - z mojego punktu widzenia transakcyjność o której piszesz to feature Rabbita, a nie niedołężność Pub/Sub. Jak pisałem funkcję, to automatycznie pomyślałem o obsłudze błędu w Firestore, tak jakbym to zrobił robiąc Restowego CRUD'a . Czy chciał”yś żebym w jakiś sposób zwrócił na to uwagę we wpisie na blogu i coś o tym dopisał?